### ISSUE TITLE
External sources should not be accesed outside of RMV::Derive

### ISSUE THREAD

**Issue by grimmmyshini** [vgvassilev/clad#448]
In the `~ReverseModeVistior()` we do the following:

```
m_ExternalSource->ForgetRMV();
``` 
which goes ahead and accesses the non-owning `m_Sources`. This is not correct because you have no guarantee that the sources were not deleted before the RMV object went out of scope. Here, you have two options:

1) Make the sources use `shared_ptr`s instead. 
2) Do this sort of cleanup after `RMV::Derive` ends.

**Comment by parth-07** [comment]
I want to understand the issue in more detail. 

> which goes ahead and accesses the non-owning m_Sources.

Please correct me if I am wrong, but isn't `m_ExternalSource` the exclusive owner of `m_Sources`?

> This is not correct because you have no guarantee that the sources were not deleted before the RMV object went out of scope.

`MultiplexExternalRMVSource` does not provide any interface to delete `ExternalSource`s. Thus, if I am not missing anything, they should not be deleted before the RMV object went out of scope. 

You mentioned one way to resolve this is: 

> Do this sort of cleanup after RMV::Derive ends.

`~ReverseModeVisitor()` would always be called after `ReverseModeVisitor::Derive` ends, thus clean up is already happening after `Derive` ends.

**Comment by grimmmyshini** [comment]
> I want to understand the issue in more detail.
> 
> > which goes ahead and accesses the non-owning m_Sources.
> 
> Please correct me if I am wrong, but isn't `m_ExternalSource` the exclusive owner of `m_Sources`?

No, I am talking about the pointers that m_Sources holds. Those are sll raw pointers that don't signify any kind of ownership. What they mean is that they are likely owned by someone else. 

> > This is not correct because you have no guarantee that the sources were not deleted before the RMV object went out of scope.
> 
> `MultiplexExternalRMVSource` does not provide any interface to delete `ExternalSource`s. Thus, if I am not missing anything, they should not be deleted before the RMV object went out of scope.

The person who does own the ExternalSource objects registered with the plugin system may delete those. That leaves you with dangling references. Since there is no way you signify ownership, getting to behaviour like this is very easy. 

> You mentioned one way to resolve this is:
> 
> > Do this sort of cleanup after RMV::Derive ends.
> 
> `~ReverseModeVisitor()` would always be called after `ReverseModeVisitor::Derive` ends, thus clean up is already happening after `Derive` ends.

No what I meant is do it inside derive, once it's about to end.

**Comment by grimmmyshini** [comment]
@parth-07 ping on this.

**Comment by parth-07** [comment]
Sorry, I did not see your last comment before. 

> No, I am talking about the pointers that m_Sources holds. Those are sll raw pointers that don't signify any kind of ownership. What they mean is that they are likely owned by someone else.

Oh, now I understand the problem. Yes, the sources does not signify any kind of ownership. Calling `m_ExternalSource->ForgetRMV();` after `ReverseModeVisitor::Derive` ends can be risky if any of the sources has been deleted. 

> The person who does own the ExternalSource objects registered with the plugin system may delete those. That leaves you with dangling references. Since there is no way you signify ownership, getting to behaviour like this is very easy.

Yes, I totally agree. What do you think about changing sources type from raw pointer to `unique_ptr`? This will solve the ownership problem.

**Comment by parth-07** [comment]
> What do you think about changing sources type from raw pointer to unique_ptr? This will solve the ownership problem.

And of course, any of the solutions that you suggested in the issue description would work as well.

**Comment by grimmmyshini** [comment]
I would suggest anything other than `uniqiue_ptr` because technically that pointer is shared between the `m_Sources` and the owner themselves. I suppose the best bet is to just make them `shared_ptr`.

**Comment by parth-07** [comment]
> I would suggest anything other than uniqiue_ptr because technically that pointer is shared between the m_Sources and the owner themselves.

`unique_ptr` approach, by means of move semantics, would typically ensure that the source pointer would not be shared. 

> I suppose the best bet is to just make them shared_ptr

`shared_ptr` has significant overhead over `unique_ptr`, thus we should use `shared_ptr` only if we really require shared pointer functionality. 

Please correct me if I am wrong, we are creating a new source for each instance of `ReverseModeVisitor`, right?

**Comment by grimmmyshini** [comment]
The whole point of this question is that `m_Sources` does not own the object it holds. Hence, make no sense to use unique_ptr here because that then implies that `m_Sources` owns those objects.  so the correct way to model this would be shared_ptrs. If you are concerned about the speed, then simply do the cleanup before derive ends, or not do the cleanup at all.

**Comment by grimmmyshini** [comment]
Honestly, I am not fully convinced by this but we can give it a try. Can you put in a PR for the changes you have in mind?

**Comment by parth-07** [comment]
We should first discuss and decide what design we want for the `MultiplexExternalRMVSource`. We have a few options: 

- Design `MultiplexExternalRMVSource` to be an exclusive owner of all external sources it holds. We can enforce this ownership using `std::unique_ptr` and move semantics.
- Design `MultiplexExternalRMVSource` to have shared ownership of all external sources it holds. We can enforce this ownership using `std::shared_ptr`.
- Follow the approach used by [`clang::MultiplexExternalSemaSource`](https://clang.llvm.org/doxygen/classclang_1_1MultiplexExternalSemaSource.html). This approach designs multiplex external source to have no ownership of the external sources it holds. In this design, it is the responsibility of a developer using the plugin system to ensure that the external sources have lifetime for the entire lifecycle of the corresponding `ReverseModeVisitor` object.

We are currently following the design used by Clang's `MultiplexExternalSemaSource`. Which design would you suggest fit best for Clad use case?

**Comment by grimmmyshini** [comment]
Honestly, I don't know. Maybe @vgvassilev can comment on this?
